### ISSUE TITLE
Allow the use of C++11 features in our compiler

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#10525]
I just got bit (again) by using a C++11-only function when writing new compiler code.  There are some nice features there, it'd be good to actually be able to use them.

**Comment by dlongnecke-cray** [comment]
I'd like to second this. I have a certain expectation for C++11 features (out of habit).

While working on bringing the multi-locale library prototype to master, I got bit several times by Travis for using C++11 features. My local `clang++` uses C++11 (even when CHPL_DEVELOPER=1) while Travis seems to use C++03.

With the switch to LLVM rapidly approaching, it seems like this should be an easy thing to change. @ronawho tells me that LLVM requires C++11 in order to build.

**Comment by bradcray** [comment]
I think this got labeled with the Chapel 2.0 epic incorrectly, probably due to an (understandable) mental mix-up with the compiler 2.0 effort.

**Comment by mppf** [comment]
The compiler has a lot of mis-uses of `virtual` for functions that are not overridden in subclasses. I'd like to start requiring C++11 so that we can use the `override` keyword as part of cleaning this up.

**Comment by leekillough** [comment]
C++11 is a giant leap from C++03. Variadic templates, `constexpr` functions and variables, `auto` variables, rvalue references, multithreading, and lambdas, not to mention a whole rewrite of the STL.

But I would use `-std=c++14` rather than `-std=c++11` since C++14 is a "bug fix release" which fixes things like relaxing `constexpr` functions, adding variable templates, adding the `auto` return type, adding `decltype(auto)`, adding generic lambdas, allowing lambda move captures, adding `std::make_unique`, adding `std::something_t` traits like `std::enable_if_t`, adding `std::index_sequence` for the non-recursive traversal of tuples or variadic arguments, etc. C++14 is what C++11 _should have been_.

C++17 is a giant leap from C++14, with its automatic deduction of class template parameters from constructor arguments, its fold expressions, `inline` variables, `constexpr` lambda expressions, `if constexpr`, structured bindings, `std::string_view`, `std::optional`, `std::variant`, `std::something_v` traits, etc. I don't think we need to jump all the way to C++17 (much less C++20) yet.

**Comment by mppf** [comment]
Let's get to C++11 before we worry about C++14. Baby steps / perfect enemy of the good / all that.

**Comment by leekillough** [comment]
> Let's get to C++11 before we worry about C++14. Baby steps / don't let perfect be the enemy of the good / all that.

I disagree. C++14 was a very important bugfix update, and has been supported in compilers for years. Just having `auto` return type, less restricted `constexpr` functions, variable templates, generic lambdas, `std::make_unique`, and `std::some_trait_t<>` are enough to justify C++14 instead of C++11. GCC and Clang have long been C++14-compliant.

We might as well jump to C++14, use it and test with it, rather than be limited to C++11 and then have to go back later and upgrade it to C++14. "Skip every other film sequel."

**Comment by lydia-duncan** [comment]
I think there's definitely an argument to be made for "2011 was 10 years ago, let's go with something a *little* more recent than that since we're upgrading"

**Comment by dlongnecke-cray** [comment]
> Let's get to C++11 before we worry about C++14. Baby steps / don't let perfect be the enemy of the good / all that.

I am increasingly also a fan of "let's get things going now and worry about perfection later", but in this case my concern is: how often will we revisit/upgrade our C++ standard? If this is a one-and-done sort of thing where the next time we explore C++30 is with Chapel 3.0, then I'd be more inclined to jump right to C++14 now while we can.

However if we consider being more aggressive about upgrading our C++ standard now that we're leaning heavily on LLVM, I'm OK with just upgrading to C++11 now as I don't think anyone on the team will argue against it.

**Comment by mppf** [comment]
We don't generally request a C++ standard when we are compiling the compiler. If the compiler is new enough that the default is > C++11 we just use that.

> how often will we revisit/upgrade our C++ standard

Our code compiles with C++03 *and newer* today. So it's not taking effort to "upgrade". It's merely a question of what we are going to require in terms of C++ compilers to build Chapel. E.g. what is the minimum gcc/clang/whatever requirement is on systems that want to build our releases? I know C++11 is OK for that but I do not know if C++14 is.

> However if we consider being more aggressive about upgrading our C++ standard now that we're leaning heavily on LLVM

Once LLVM is the default, we will be bundling a recent clang compiler and expecting people to build it / have it installed. This makes it easier IMO to require newer C++ because we include a working compiler. However we'd have to change to building the compiler with the selected `clang` for that story to work out.

**Comment by mppf** [comment]
I've learned recently that as of LLVM 9, LLVM requires C++14. At the same time, we are switching to LLVM by default. As a result, I'd like to now propose that we require C++14 to build the compiler soon. 

In particular, for the near term:
 * update the prerequisites page to indicate C++14 is required
 * require C++14 in the new compiler revamp code and use its features now (it will make GCC better at some of the variadic templates; but doesn't have much near-term risk since this code isn't yet linked in to the normal compiler build)
 * start using C++14 features as needed in the main compiler only after we have achieved LLVM-by-default.
